Assert that GIL is not held in blocking destructors

ezyang opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#57030 Assert that GIL is not held in blocking destructors**
* #57029 Release GIL before destructing RPCAgent subclasses.

PR #57029 is not perfect; there are still obscure situations in which
we might allocate a shared_ptr to an RpcAgent that doesn't have a
no GIL constructor, so this PR adds the other half of the equation:
assert that we don't hold the GIL when running a blocking destructor.
This makes it possible to detect potential deadlocks even if the
code doesn't deadlock in practice (because you got lucky and none
of the threads you blocked on tried to also take out the GIL).

I considered whether or not to make this DEBUG_ONLY.  For now it's
not, so I can get better CI coverage, and because this test only
happens in destructors of objects that die rarely.

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

Differential Revision: [D28030582](https://our.internmc.facebook.com/intern/diff/D28030582)

ezyang committed:
Assert that GIL is not held in blocking destructors

PR #57029 is not perfect; there are still obscure situations in which
we might allocate a shared_ptr to an RpcAgent that doesn't have a
no GIL constructor, so this PR adds the other half of the equation:
assert that we don't hold the GIL when running a blocking destructor.
This makes it possible to detect potential deadlocks even if the
code doesn't deadlock in practice (because you got lucky and none
of the threads you blocked on tried to also take out the GIL).

I considered whether or not to make this DEBUG_ONLY.  For now it's
not, so I can get better CI coverage, and because this test only
happens in destructors of objects that die rarely.

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

[ghstack-poisoned]

ezyang committed:
Update on "Assert that GIL is not held in blocking destructors"


PR #57029 is not perfect; there are still obscure situations in which
we might allocate a shared_ptr to an RpcAgent that doesn't have a
no GIL constructor, so this PR adds the other half of the equation:
assert that we don't hold the GIL when running a blocking destructor.
This makes it possible to detect potential deadlocks even if the
code doesn't deadlock in practice (because you got lucky and none
of the threads you blocked on tried to also take out the GIL).

I considered whether or not to make this DEBUG_ONLY.  For now it's
not, so I can get better CI coverage, and because this test only
happens in destructors of objects that die rarely.

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

Differential Revision: [D28030582](https://our.internmc.facebook.com/intern/diff/D28030582)

[ghstack-poisoned]

ezyang committed:
Update on "Assert that GIL is not held in blocking destructors"


PR #57029 is not perfect; there are still obscure situations in which
we might allocate a shared_ptr to an RpcAgent that doesn't have a
no GIL constructor, so this PR adds the other half of the equation:
assert that we don't hold the GIL when running a blocking destructor.
This makes it possible to detect potential deadlocks even if the
code doesn't deadlock in practice (because you got lucky and none
of the threads you blocked on tried to also take out the GIL).

I considered whether or not to make this DEBUG_ONLY.  For now it's
not, so I can get better CI coverage, and because this test only
happens in destructors of objects that die rarely.

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

Differential Revision: [D28030582](https://our.internmc.facebook.com/intern/diff/D28030582)

[ghstack-poisoned]

albanD commented on c10/util/DeadlockDetection.cpp:
```diff
@@ -0,0 +1,21 @@
+PythonGILHooks* python_gil_hooks = nullptr;
+}
+
+bool check_python_gil() {
```
nit: "check" does not give any information if we expect it to be held or not.
Maybe `get_python_gil_state()`?

ezyang replied:
I'm drawing analogy with the analogous CPython API `PyGILState_Check`. BTW, "get" is bad because it aliases with similar sounding `PyGILState_GetThisThreadState` and `PyThreadState_Get`

ezyang replied:
Also the name here doesn't really matter, it's not public API

albanD replied:
Similarity with `PyGILState_Check` sounds good :)

albanD commented on c10/util/DeadlockDetection.h:
```diff
@@ -0,0 +1,48 @@
+namespace c10 {
+
+#define TORCH_ASSERT_NO_GIL_WITHOUT_PYTHON_DEP() \
+  TORCH_INTERNAL_ASSERT(!c10::impl::check_python_gil())
```
Do we want to add an special message here?

albanD commented on c10/util/DeadlockDetection.h:
```diff
@@ -0,0 +1,48 @@
+C10_API bool check_python_gil();
+
+struct C10_API PythonGILHooks {
+  virtual ~PythonGILHooks() = default;
```
Out of curiosity, why do you need to specify this destructor explicitly?

ezyang replied:
By default destructor is not virtual. https://stackoverflow.com/questions/461203/when-to-use-virtual-destructors

Technically it won't matter here because I will never delete from a base pointer but it's good hygiene.

albanD reviewed (approved):
LGTM
Just small questions

ezyang committed:
Update on "Assert that GIL is not held in blocking destructors"


PR #57029 is not perfect; there are still obscure situations in which
we might allocate a shared_ptr to an RpcAgent that doesn't have a
no GIL constructor, so this PR adds the other half of the equation:
assert that we don't hold the GIL when running a blocking destructor.
This makes it possible to detect potential deadlocks even if the
code doesn't deadlock in practice (because you got lucky and none
of the threads you blocked on tried to also take out the GIL).

I considered whether or not to make this DEBUG_ONLY.  For now it's
not, so I can get better CI coverage, and because this test only
happens in destructors of objects that die rarely.

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

Differential Revision: [D28030582](https://our.internmc.facebook.com/intern/diff/D28030582)

[ghstack-poisoned]
